src/hotspot/share/runtime/threads.cpp

```diff
@@ -1331,9 +1331,11 @@ void Threads::print_on(outputStream* st, bool print_stacks,
         if (thread_oop != nullptr) {
           if (p->is_vthread_mounted()) {
             const oop vt = p->vthread();
-            assert(vt != nullptr, "vthread should not be null when vthread is mounted");
-            st->print_cr("   Mounted virtual thread \"%s\" #" INT64_FORMAT, JavaThread::name_for(vt), (int64_t)java_lang_Thread::thread_id(vt));
-            p->print_vthread_stack_on(st);
+            if (vt != thread_oop) {
+              assert(vt != nullptr, "vthread should not be null when vthread is mounted");
```

Comment:
I think the assert still belongs in the original position doesn't it? Or could the problem we hit here cause a transient null to appear as well?

Reply:
I think that, even if based on [this](https://github.com/openjdk/jdk/pull/19482#issuecomment-2166116062) comment from @AlanBateman's in some cases the virtual thread can be mounted but `_vthread` can refer to the carrier thread, what you're saying is correct and in both cases `_vthread` should still not be a null pointer:

- Either it's equal to `thread_oop` and we've already checked that it's not `nullptr`.
- Or it's different to `thread_oop` and it refers to an actual virtual thread that should not be null.

Reply:
Moving the assert before the `if` in: 9a954efe164d9caee31494773c0eecc8ff69223f